### PR TITLE
test(a11y): playwright coverage for explorer - search page - bed-8633

### PR DESCRIPTION
**Pull request by JayC-Specter**
## Description

Adds accessibility tests for the Explore Search page to verify compliance with accessibility standards.

## Motivation and Context

Resolves bed-8633

This change was needed to ensure the Explore Search page adheres to accessibility standards.

## How Has This Been Tested?

The tests were run locally.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist:

- [x ] I have met the contributing prerequisites
  - Assigned myself to this PR
  - Added the appropriate labelsx
  - Associated an issue: https://github.com/SpecterOps/BloodHound/issues/672
  - Read the Contributing guide: https://github.com/SpecterOps/BloodHound/wiki/Contributing
- [x ] I have ensured that related documentation is up-to-date
  - Open API docs
  - Code comments (GoDocs / JSDocs)
- [x ] I have followed proper test practices
  - Added/updated tests to cover my changes
  - All new and existing tests passed

## Summary by CodeRabbit

* **Tests**
  * Added a Playwright accessibility spec for the Explore page.
  * Verified search tab behavior, including focus and the search prompt.
  * Added coverage for search results and no-results states, including visible UI messaging.
  * Included WCAG A/AA accessibility checks for the scoped search experience.

### CODE REVIEW

**Comment by JayC-Specter** [comment]
I have read the CLA Document and I hereby sign the CLA

File: cmd/ui/tests/a11y/explore-search.a11y.spec.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+import { expect, expectNoAccessibilityViolations, test } from '../fixtures';
+
+test.describe('Explore page accessibility', () => {
+    test('focuses the node search field on navigation', async ({ page }) => {
+        await page.goto('/ui/explore');
```
Comment (TheNando) [inline cmd/ui/tests/a11y/explore-search.a11y.spec.ts:20]: Since each test is for the same page, you could add a single `test.beforeEach` at the top, just below the describe block, to do the page visit step. Then you don't need to explicitly have the goto in each test.

```
    test.beforeEach('setup', async ({ page }) => {
        await page.goto('/ui/explore');
    });
```

File: cmd/ui/tests/a11y/explore-search.a11y.spec.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+        await expect(page.getByLabel('Search Nodes')).toBeFocused();
```
Comment (TheNando) [inline cmd/ui/tests/a11y/explore-search.a11y.spec.ts]: You can replace this assertion with a basic wait
```
await page.getByText('Begin typing to search').waitFor({ state: 'visible' });
```

And add the axe assessment to the bottom. This should be at the end of every test, though the `.include` may change depending on the circumstance.
```
const results = await makeAxeBuilder().include('#content-wrapper').analyze();
await expectNoAccessibilityViolations(testInfo, results, { page });
```

However, for this test, there is an extra gotcha that I don't yet know what to do about. If we intentionally add a contrast issue to the search instruction popover, this test will still pass, missing the contrast issue. If you dig through the report, you'll find the contrast issue reported as "incomplete". This is maybe a bug or edge case where for some reason, Axe-core's logic thinks something is covering the content so it can't determine if the contrast issue is valid or not. We'll just have to accept this for now and either manually track these or see what else we come up with.

File: cmd/ui/tests/a11y/explore-search.a11y.spec.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+import { expect, expectNoAccessibilityViolations, test } from '../fixtures';
+
+test.describe('Explore page accessibility', () => {
```
Comment (TheNando) [inline cmd/ui/tests/a11y/explore-search.a11y.spec.ts]: The early tests we wrote have a lot of redundancy. You can move the mention of the WCAG stuff to the top, then you don't need to repeat it in the test titles.
```suggestion
test.describe('WCAG A/AA violations - Explore - Search Tab', () => {
```

File: cmd/ui/tests/a11y/explore-search.a11y.spec.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    test('focuses the node search field on navigation', async ({ page }) => {
```
Comment (TheNando) [inline cmd/ui/tests/a11y/explore-search.a11y.spec.ts]: Because these are accessibility tests, we don't need to make assertions about state. We just want to wait until the page or state is loaded such that elements are visible then run the axe assessment test. So the test titles can just be the various scenarios and states which show differing content.

```suggestion
    test('Search tab', async ({ page }) => {
```

File: cmd/ui/tests/a11y/explore-search.a11y.spec.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        await page.goto('/ui/explore');
+
+        await expect(page.getByLabel('Search Nodes')).toBeFocused();
+    });
+
+    test('node search results state has no detectable WCAG A/AA violations', async ({
```
Comment (TheNando) [inline cmd/ui/tests/a11y/explore-search.a11y.spec.ts]:
```suggestion
    test('Search with results', async ({
```

File: cmd/ui/tests/a11y/explore-search.a11y.spec.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            if (route.request().method() !== 'GET') {
+                return route.fallback();
+            }
+            await route.fulfill({
+                json: {
+                    data: [{ name: searchResultName, objectid: 'playwright-search-result', type: 'User' }],
+                },
+            });
+        });
+        await page.goto('/ui/explore');
```
Comment (TheNando) [inline cmd/ui/tests/a11y/explore-search.a11y.spec.ts]: You wont need this line if you add the `beforeEach`

File: cmd/ui/tests/a11y/explore-search.a11y.spec.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        const searchField = page.getByLabel('Search Nodes');
+        await searchField.fill(searchTerm);
+        const searchResult = page.getByRole('option').filter({ hasText: searchResultName });
+
+        await expect(searchResult).toBeVisible();
+        const results = await makeAxeBuilder().include('#content-wrapper').analyze();
+        await expectNoAccessibilityViolations(testInfo, results, { page });
+    });
+
+    test('node search no-results state has no detectable WCAG A/AA violations', async ({
```
Comment (TheNando) [inline cmd/ui/tests/a11y/explore-search.a11y.spec.ts]:
```suggestion
    test('Search with no results', async ({
```

File: cmd/ui/tests/a11y/explore-search.a11y.spec.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+        await page.route('**/api/v2/search**', async (route) => {
+            if (route.request().method() !== 'GET') {
+                return route.fallback();
+            }
+
+            await route.fulfill({ json: { data: [] } });
+        });
+
+        await page.goto('/ui/explore');
```
Comment (TheNando) [inline cmd/ui/tests/a11y/explore-search.a11y.spec.ts]: Removed.

**Review by TheNando — COMMENTED** [review]
Overall, you nailed the big ideas but I have some more info about the specific details.

**Review by TheNando — APPROVED** [review]

File: cmd/ui/tests/a11y/explore-search.a11y.spec.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+test.describe('WCAG A/AA Violations - Explore - Search Tab', () => {
+    test.beforeEach(async ({ page }) => {
+        await page.goto('/ui/explore');
+    });
+
+    test('Search tab', async ({ page, makeAxeBuilder }, testInfo) => {
+        await page.getByText('Begin typing to search').waitFor({ state: 'visible' });
+
+        const results = await makeAxeBuilder().include('#content-wrapper').analyze();
+        await expectNoAccessibilityViolations(testInfo, results, { page });
```
[the remark this answers is not in the document]
Comment (TheNando) [reply cmd/ui/tests/a11y/explore-search.a11y.spec.ts:27]: Tests with `.a11y.` in the filename are accessibility tests. The purpose of each test is to get the UI into a unique state and run the axe builder analyze then check the results for violations. There is no need to make explicit assertions outside of that for accessibility tests.
